File: 프로그래머스/L3/등산코스 정하기/등산코스 정하기.js
function solution(n, paths, gates, summits) {
  const graph = new Array(n + 1).fill(null).map((_) => []);
  for (let i = 0; i < paths.length; i++) {
    const [a, b, w] = paths[i];
    graph[a].push([w, b]);
    graph[b].push([w, a]);
  }
  for (let summit of summits) {
    graph[summit] = [];
  }
  let q = gates;
  const dp = new Array(n + 1).fill(Infinity);
  gates.forEach((v) => (dp[v] = -1));
  while (q.length > 0) {
    let set = new Set();
    while (q.length > 0) {
      const qv = q.pop();
      for (let [w, v] of graph[qv]) {
        const maxV = Math.max(dp[qv], w);
        if (dp[v] > maxV) {
          dp[v] = maxV;
          set.add(v);
        }
      }
    }
    q = [...set];
  }

  const res = summits
    .map((v) => [v, dp[v]])
    .sort((a, b) => {
      if (a[1] === b[1]) {
        return a[0] - b[0];
      }
      return a[1] - b[1];
    });
  return res[0];
}

File: 프로그래머스/L3/등산코스 정하기/등산코스 정하기.py
import heapq


def solution(n, paths, gates, summits):
    answer = []
    graph = [[] for i in range(n+1)]
    for a, b, w in paths:
        graph[b].append([w, a])
        graph[a].append([w, b])

    for summit in summits:
        graph[summit] = []

    def dijkstra():
        summitsSet = set(summits)
        dp = [float('inf') for i in range(n+1)]
        for gate in gates:
            dp[gate] = 0

        hq = [[0, gate] for gate in gates]
        heapq.heapify(hq)
        while(hq):
            hqW, hqV = heapq.heappop(hq)

            if hqV in summitsSet or hqW > dp[hqV]:
                continue

            for [gW, gV] in graph[hqV]:
                m = max(gW, dp[hqV])
                if(dp[gV] > m):
                    dp[gV] = m
                    heapq.heappush(hq, [m, gV])

        return dp
    dp = dijkstra()
    summits.sort(key=lambda x: (dp[x], x))
    return [summits[0], dp[summits[0]]]
